code/modules/fabrication/_fabricator.dm

```diff
@@ -0,0 +1,279 @@
+			recipe_data["can_make"] = !isnull(stored_material[resource]) && stored_material[resource] < round(R.resources[resource]*mat_efficiency)
+		recipe_data["category"] = R.category
+		recipe_data["resources"] = english_list(resources)
+		recipe_data["build_time"] = (R.build_time / 10)
+		recipe_data["max_sheets"] = null
+		if(R.is_stack)
+			var/obj/item/stack/R_stack = R.path
+			recipe_data["max_sheets"] = initial(R_stack.max_amount)
+		data["recipes"] += list(recipe_data)
+
+	data["currently_printing"] = null
+	if(currently_printing)
+		data["currently_printing"] = "\ref[currently_printing]"
+	data["queue"] = list()
+	for(var/datum/fabricator_build_order/AR in print_queue)
+		data["queue"] += list(
+			list(
+				"ref" = "\ref[AR]",
+				"order" = AR.target_recipe.name,
+				"path" = AR.target_recipe.type,
+				"multiplier" = AR.multiplier,
+				"build_time" = AR.target_recipe.build_time,
+				"progress" = AR.target_recipe.build_time - AR.remaining_time,
+				"remaining_time" = AR.remaining_time
+			)
+		)
+	return data
+
+/obj/machinery/fabricator/attackby(obj/item/attacking_item, mob/user)
+	if(fab_status_flags & FAB_BUSY)
+		to_chat(user, SPAN_NOTICE("\The [src] is busy. Please wait for the completion of previous operation."))
+		return TRUE
+
+	if(default_deconstruction_screwdriver(user, attacking_item))
+		SStgui.update_uis(src)
+		return TRUE
+	if(default_deconstruction_crowbar(user, attacking_item))
+		return TRUE
+	if(default_part_replacement(user, attacking_item))
+		return TRUE
+
+	if(stat)
+		return TRUE
+
+	if(panel_open)
+		//Don't eat multitools or wirecutters used on an open lathe.
+		if(attacking_item.ismultitool() || attacking_item.iswirecutter())
+			if(panel_open)
+				wires.interact(user)
+			else
+				to_chat(user, SPAN_WARNING("\The [src]'s wires aren't exposed."))
+			return TRUE
+
+	if(attacking_item.loc != user && !istype(attacking_item, /obj/item/stack))
+		return FALSE
+
+	if(is_robot_module(attacking_item))
+		return FALSE
+
+	load_lathe(attacking_item, user)
+	return TRUE
+
+/obj/machinery/fabricator/attack_hand(mob/user)
+	user.set_machine(src)
+	ui_interact(user)
+
+/obj/machinery/fabricator/proc/is_functioning()
+	. = use_power != POWER_USE_OFF && !(fab_status_flags & FAB_DISABLED)
+
+/obj/machinery/fabricator/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+
+	usr.set_machine(src)
+	add_fingerprint(usr)
+
+	playsound(src, /singleton/sound_category/keyboard_sound, 50)
+
+	if(action == "make")
+		var/multiplier = text2num(params["multiplier"])
+		var/singleton/fabricator_recipe/R = GET_SINGLETON(text2path(params["recipe"]))
+		if(!istype(R))
+			CRASH("Unknown recipe given! [R], param is [params["recipe"]].")
+
+		intent_message(MACHINE_SOUND)
+
+		try_queue_build(R, multiplier)
+
+		. = TRUE
+
+	if(action == "remove")
+		var/datum/fabricator_build_order/order = locate(params["ref"])
+		try_cancel_build(order)
+		qdel(order)
```

Comment:
Unguarded qdel, this can be exploited to delete anything that the caller can find the ref of, including mobs, subsystems and what have you